File: src/Axol.Lexer/Lexer.cs
using Axol.Core;
using Axol.Core.Tokens;

namespace Axol.Lexer;

public sealed class AxolLexer
{
    private readonly string _source;
    private readonly string _file;
    private int _pos;

    public string Source => _source;

    public AxolLexer(string source, string file = "<stdin>")
    {
        _source = source;
        _file = file;
        _pos = 0;
    }

    public List<Token> Tokenize()
    {
        var tokens = new List<Token>();
        while (true)
        {
            var tok = NextToken();
            tokens.Add(tok);
            if (tok.Kind == TokenKind.Eof || tok.Kind == TokenKind.Error)
                break;
        }
        return tokens;
    }

    public Token NextToken()
    {
        SkipWhitespaceAndComments();

        if (_pos >= _source.Length)
            return Token.Eof(Span(_pos, _pos));

        var start = _pos;
        var ch = _source[_pos];

        switch (ch)
        {
            case '(':
                _pos++;
                return new Token(TokenKind.LParen, "(", null, Span(start, _pos));
            case ')':
                _pos++;
                return new Token(TokenKind.RParen, ")", null, Span(start, _pos));
            case '[':
                _pos++;
                return new Token(TokenKind.LBracket, "[", null, Span(start, _pos));
            case ']':
                _pos++;
                return new Token(TokenKind.RBracket, "]", null, Span(start, _pos));
            case '{':
                _pos++;
                return new Token(TokenKind.LBrace, "{", null, Span(start, _pos));
            case '}':
                _pos++;
                return new Token(TokenKind.RBrace, "}", null, Span(start, _pos));
            case '"':
                return ReadString();
            case '-' when _pos + 1 < _source.Length && _source[_pos + 1] == '>':
                _pos += 2;
                return new Token(TokenKind.Arrow, "->", null, Span(start, _pos));
            default:
                if (ch == '-' && _pos + 1 < _source.Length && char.IsDigit(_source[_pos + 1]))
                    return ReadNumber();
                if (char.IsDigit(ch))
                    return ReadNumber();
                return ReadSymbol();
        }
    }

    private Token ReadString()
    {
        var start = _pos;
        _pos++; // skip opening "
        var sb = new System.Text.StringBuilder();

        while (_pos < _source.Length && _source[_pos] != '"')
        {
            if (_source[_pos] == '\\' && _pos + 1 < _source.Length)
            {
                _pos++;
                switch (_source[_pos])
                {
                    case 'n': sb.Append('\n'); break;
                    case 't': sb.Append('\t'); break;
                    case '\\': sb.Append('\\'); break;
                    case '"': sb.Append('"'); break;
                    default: sb.Append('\\'); sb.Append(_source[_pos]); break;
                }
                _pos++;
            }
            else
            {
                sb.Append(_source[_pos]);
                _pos++;
            }
        }

        if (_pos >= _source.Length)
            return Token.Error("Unterminated string", Span(start, _pos));

        _pos++; // skip closing "
        var lexeme = _source[start.._pos];
        return new Token(TokenKind.StrLiteral, lexeme, sb.ToString(), Span(start, _pos));
    }

    private Token ReadNumber()
    {
        var start = _pos;
        if (_source[_pos] == '-') _pos++;

        while (_pos < _source.Length && char.IsDigit(_source[_pos]))
            _pos++;

        bool isFloat = false;
        if (_pos < _source.Length && _source[_pos] == '.' && _pos + 1 < _source.Length && char.IsDigit(_source[_pos + 1]))
        {
            isFloat = true;
            _pos++; // skip dot
            while (_pos < _source.Length && char.IsDigit(_source[_pos]))
                _pos++;
        }

        var lexeme = _source[start.._pos];

        if (isFloat)
        {
            var val = double.Parse(lexeme, System.Globalization.CultureInfo.InvariantCulture);
            return new Token(TokenKind.FloatLiteral, lexeme, val, Span(start, _pos));
        }
        else
        {
            var val = long.Parse(lexeme);
            return new Token(TokenKind.IntLiteral, lexeme, val, Span(start, _pos));
        }
    }

    private Token ReadSymbol()
    {
        var start = _pos;
        while (_pos < _source.Length && !IsDelimiter(_source[_pos]))
            _pos++;

        var lexeme = _source[start.._pos];

        // Handle "..." rest patterns
        if (lexeme.EndsWith("..."))
        {
            return new Token(TokenKind.Symbol, lexeme, null, Span(start, _pos));
        }

        if (lexeme == "true")
            return new Token(TokenKind.Symbol, lexeme, true, Span(start, _pos));
        if (lexeme == "false")
            return new Token(TokenKind.Symbol, lexeme, false, Span(start, _pos));
        if (lexeme == "nil")
            return new Token(TokenKind.Symbol, lexeme, null, Span(start, _pos));

        return new Token(TokenKind.Symbol, lexeme, null, Span(start, _pos));
    }

    private void SkipWhitespaceAndComments()
    {
        while (_pos < _source.Length)
        {
            if (char.IsWhiteSpace(_source[_pos]))
            {
                _pos++;
            }
            else if (_source[_pos] == ';')
            {
                // Skip line comment
                while (_pos < _source.Length && _source[_pos] != '\n')
                    _pos++;
            }
            else
            {
                break;
            }
        }
    }

    private static bool IsDelimiter(char ch) =>
        char.IsWhiteSpace(ch) || ch == '(' || ch == ')' || ch == '[' || ch == ']'
        || ch == '{' || ch == '}' || ch == '"' || ch == ';';

    private SourceSpan Span(int start, int end) => new(_file, start, end);
}

File: axol-lang/src/dsl/lexer.rs
//! AXOL DSL Lexer — tokenizes .axol source files.
//!
//! Syntax example:
//! ```axol
//! declare "mood_classifier" {
//!     input text_vec(128)
//!     input context(64)
//!     relate sentiment <- text_vec, context via <~>
//!     output sentiment
//!     quality omega=0.9 phi=0.8
//! }
//!
//! weave mood_classifier quantum=true seed=42
//! observe mood_classifier { text_vec = [0.1, 0.2, ...] }
//! ```

#[derive(Clone, Debug, PartialEq)]
pub enum Token {
    // Keywords
    Declare,
    Input,
    Output,
    Relate,
    Via,
    Quality,
    Weave,
    Observe,
    Reobserve,
    Quantum,
    Seed,
    Count,

    // Compose keywords
    Compose,
    Gate,
    Confident,
    Iterate,
    Converge,
    Then,
    Else,
    Design,
    Learn,

    // Literals
    Ident(String),
    StringLit(String),
    Number(f64),
    Int(i64),

    // Symbols
    LBrace,     // {
    RBrace,     // }
    LParen,     // (
    RParen,     // )
    LBracket,   // [
    RBracket,   // ]
    Arrow,      // <-
    Comma,      // ,
    Equals,     // =
    Relation(String), // <~>, <+>, <*>, <!>, <?>

    // Control
    Newline,
    Eof,
}

pub struct Lexer {
    input: Vec<char>,
    pos: usize,
    pub tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(source: &str) -> Self {
        Self {
            input: source.chars().collect(),
            pos: 0,
            tokens: Vec::new(),
        }
    }

    pub fn tokenize(&mut self) -> &[Token] {
        while self.pos < self.input.len() {
            self.skip_whitespace();
            if self.pos >= self.input.len() { break; }

            let ch = self.input[self.pos];

            match ch {
                '#' => { self.skip_line(); }
                '\n' | '\r' => {
                    self.tokens.push(Token::Newline);
                    self.pos += 1;
                    if ch == '\r' && self.peek() == Some('\n') { self.pos += 1; }
                }
                '{' => { self.tokens.push(Token::LBrace); self.pos += 1; }
                '}' => { self.tokens.push(Token::RBrace); self.pos += 1; }
                '(' => { self.tokens.push(Token::LParen); self.pos += 1; }
                ')' => { self.tokens.push(Token::RParen); self.pos += 1; }
                '[' => { self.tokens.push(Token::LBracket); self.pos += 1; }
                ']' => { self.tokens.push(Token::RBracket); self.pos += 1; }
                ',' => { self.tokens.push(Token::Comma); self.pos += 1; }
                '=' => { self.tokens.push(Token::Equals); self.pos += 1; }
                '<' => { self.read_angle(); }
                '"' => { self.read_string(); }
                _ if ch.is_ascii_digit() || (ch == '-' && self.peek_next_digit()) => {
                    self.read_number();
                }
                _ if ch.is_ascii_alphabetic() || ch == '_' => {
                    self.read_ident();
                }
                _ => { self.pos += 1; } // skip unknown
            }
        }
        self.tokens.push(Token::Eof);
        &self.tokens
    }

    fn skip_whitespace(&mut self) {
        while self.pos < self.input.len() && self.input[self.pos] == ' ' || self.pos < self.input.len() && self.input[self.pos] == '\t' {
            self.pos += 1;
        }
    }

    fn skip_line(&mut self) {
        while self.pos < self.input.len() && self.input[self.pos] != '\n' {
            self.pos += 1;
        }
    }

    fn peek(&self) -> Option<char> {
        self.input.get(self.pos).copied()
    }

    fn peek_next_digit(&self) -> bool {
        self.input.get(self.pos + 1).map_or(false, |c| c.is_ascii_digit())
    }

    fn read_angle(&mut self) {
        // Could be <-, <~>, <+>, <*>, <!>, <?>
        let start = self.pos;
        self.pos += 1; // skip '<'

        if self.pos < self.input.len() && self.input[self.pos] == '-' {
            self.pos += 1;
            self.tokens.push(Token::Arrow);
            return;
        }

        // Read operator like ~>, +>, *>, !>, ?>
        let mut op = String::from("<");
        while self.pos < self.input.len() && self.input[self.pos] != '>' {
            op.push(self.input[self.pos]);
            self.pos += 1;
        }
        if self.pos < self.input.len() {
            op.push('>');
            self.pos += 1;
        }
        self.tokens.push(Token::Relation(op));
    }

    fn read_string(&mut self) {
        self.pos += 1; // skip opening "
        let mut s = String::new();
        while self.pos < self.input.len() && self.input[self.pos] != '"' {
            s.push(self.input[self.pos]);
            self.pos += 1;
        }
        if self.pos < self.input.len() { self.pos += 1; } // skip closing "
        self.tokens.push(Token::StringLit(s));
    }

    fn read_number(&mut self) {
        let start = self.pos;
        if self.input[self.pos] == '-' { self.pos += 1; }
        while self.pos < self.input.len() && self.input[self.pos].is_ascii_digit() {
            self.pos += 1;
        }
        let mut is_float = false;
        if self.pos < self.input.len() && self.input[self.pos] == '.' {
            is_float = true;
            self.pos += 1;
            while self.pos < self.input.len() && self.input[self.pos].is_ascii_digit() {
                self.pos += 1;
            }
        }
        let text: String = self.input[start..self.pos].iter().collect();
        if is_float {
            self.tokens.push(Token::Number(text.parse().unwrap_or(0.0)));
        } else {
            self.tokens.push(Token::Int(text.parse().unwrap_or(0)));
        }
    }

    fn read_ident(&mut self) {
        let start = self.pos;
        while self.pos < self.input.len() && (self.input[self.pos].is_ascii_alphanumeric() || self.input[self.pos] == '_') {
            self.pos += 1;
        }
        let word: String = self.input[start..self.pos].iter().collect();
        let token = match word.as_str() {
            "declare" => Token::Declare,
            "input" => Token::Input,
            "output" => Token::Output,
            "relate" => Token::Relate,
            "via" => Token::Via,
            "quality" => Token::Quality,
            "weave" => Token::Weave,
            "observe" => Token::Observe,
            "reobserve" => Token::Reobserve,
            "quantum" => Token::Quantum,
            "seed" => Token::Seed,
            "count" => Token::Count,
            "compose" => Token::Compose,
            "gate" => Token::Gate,
            "confident" => Token::Confident,
            "iterate" => Token::Iterate,
            "converge" => Token::Converge,
            "then" => Token::Then,
            "else" => Token::Else,
            "design" => Token::Design,
            "learn" => Token::Learn,
            "true" => Token::Int(1),
            "false" => Token::Int(0),
            _ => Token::Ident(word),
        };
        self.tokens.push(token);
    }
}
